src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -143,59 +143,82 @@ private BoundPattern BindDiscardPattern(DiscardPatternSyntax node, TypeSymbol in
+                    diagnostics.Add(ErrorCode.WRN_IsTypeNamedUnderscore, innerExpression.Location, convertedExpression.Type);
+                }
 
-            constantDiagnostics.Free();
-            return result;
+                return new BoundTypePattern(node, (BoundTypeExpression)convertedExpression, inputType, hasErrors);
+            }
         }
 
-        internal BoundConstantPattern BindConstantPattern(
-            SyntaxNode node,
+        /// <summary>
+        /// Binds the expression for a pattern.  Sets <paramref name="wasExpression"/> if it was a type rather than an expression,
+        /// and in that case it returns a <see cref="BoundTypeExpression"/>.
+        /// </summary>
+        private BoundExpression BindExpressionOrTypeForPattern(
             TypeSymbol inputType,
             ExpressionSyntax patternExpression,
             bool hasErrors,
             DiagnosticBag diagnostics,
+            out ConstantValue constantValueOpt,
             out bool wasExpression)
         {
-            var convertedExpression = BindExpressionForPattern(node, inputType, patternExpression, hasErrors, diagnostics, out var constantValueOpt, out wasExpression);
-            return new BoundConstantPattern(node, convertedExpression, constantValueOpt ?? ConstantValue.Bad, inputType, hasErrors);
+            constantValueOpt = null;
+            BoundExpression expression = BindTypeOrRValue(patternExpression, diagnostics);
+            wasExpression = expression.Kind != BoundKind.TypeExpression;
+            if (wasExpression)
+            {
+                return BindExpressionForPatternContinued(expression, inputType, patternExpression, hasErrors, diagnostics, out constantValueOpt);
+            }
+            else
+            {
+                Debug.Assert(expression.Kind == BoundKind.TypeExpression);
+                return expression;
+            }
         }
 
-        private BoundExpression BindExpressionForPattern(SyntaxNode node,
+        /// <summary>
+        /// Binds the expression for an is-type right-hand-side, in case it does not bind as a type.
+        /// </summary>
+        private BoundExpression BindExpressionForPattern(
             TypeSymbol inputType,
             ExpressionSyntax patternExpression,
             bool hasErrors,
             DiagnosticBag diagnostics,
             out ConstantValue constantValueOpt,
             out bool wasExpression)
         {
-            BoundExpression expression = BindValue(patternExpression, diagnostics, BindValueKind.RValue);
             constantValueOpt = null;
+            var expression = BindExpression(patternExpression, diagnostics: diagnostics, invoked: false, indexed: false);
+            expression = CheckValue(expression, BindValueKind.RValue, diagnostics);
+            wasExpression = expression.Kind switch { BoundKind.BadExpression => false, BoundKind.TypeExpression => false, _ => true };
```

Comment:
is `BoundKind.TypeExpression` here only observable in error scenarios?

Reply:
Yes.  When binding an is-type operator, we call this when we already tried to bind it as a type and found a problem there.  So if we get a type here we'll fall back to reporting those problems.

---
In reply to: [360573727](https://github.com/dotnet/roslyn/pull/40503#discussion_r360573727) [](ancestors = 360573727)